docs/proposals/local-registries.md

```diff
@@ -0,0 +1,165 @@
+
+First, clusters which deploy a local image registry must provide the releavant host info by
+publishing it in the `local-registry-hosting`
+[ConfigMap](https://github.com/kubernetes/enhancements/tree/master/keps/sig-cluster-lifecycle/generic/1755-communicating-a-local-registry#the-local-registry-hosting-configmap).
+The build operator/controller must have RBAC which allows it to read the `ConfigMap` and obtain the
+registry host information. For `Build` and `BuildRun` objects, the `HostFromClusterNetwork` value
+should be used as the hostname.
+
+If the `local-registry-hosting` ConfigMap is created, updated, or deleted, the build controller
+and/or operator needs to react accordingly.
+
+#### Declaring an image as local
+
+`Image` references in the API will be updated to have the `local` field. An example using the 
+`output` spec:
+
+```yaml
+spec:
+  output:
+    image: mynamespace/myapp:latest
+    local: true
+```
+
+When an image is declared as `local`, the build controller prepends the local image registry
+hostname to the provided image reference.
+
+### Risks and Mitigations
+
+App developers use `local: true` on clusters that don't have a local registry defined, end up
+pulling images from `docker.io` or an image registry in container runtime's unqualified search
+path.
+
+Local image registry configurations can be altered while the build controller is running.
+
+## Design Details
+
+### Test Plan
+
+1. Deploy the build operator/controller on a cluster that populates the `local-registry-hosting`
+   ConfigMap OR as cluster-admin, manually populate the `local-registry-hosting` ConfigMap.
+2. Create a `Build` and/or `BuildRun` which references a local image.
+3. Ensure that a build can push an image referencing the local image spec.
+
+### Graduation Criteria
+
+**Dev Preview** - we can simply define an environment variable to set the local registry hostname.
+
+**Tech Preview/GA** - fully read the `local-registry-hosting` ConfigMap, with the ability to update
+the build controller if this ConfigMap changes.
+
+#### Examples
+
+These are generalized examples to consider, in addition to the aforementioned [maturity
+levels][maturity-levels].
+
+### Upgrade / Downgrade Strategy
+
+Upgrades add a new field to the API. On downgrade the `local` attribute should be ignored.
+
+### Version Skew Strategy
+
+N/A
+
+## Implementation History
+
+2020-07-24: Initial proposal
+
+## Drawbacks
+
+TBD
+
+The idea is to find the best form of an argument why this enhancement should _not_ be implemented.
+
+## Alternatives
+
+TBD
+
+Similar to the `Drawbacks` section the `Alternatives` section is used to highlight and record other
+possible approaches to delivering the value proposed by an enhancement.
+
+## Infrastructure Needed [optional]
+
+- A cluster which populates the `local-registry-hosting` ConfigMap
+
+## Open Questions [optional]
+
+1. How can we make `Build` objects with local image pull specs portable across namespaces?
```

Comment:
The use case here is that a developer team is storing the `Build` objects in git (gitops) and is running their dev/stage/prod code in separate namespaces. Having a spec as:

```
spec:
  output:
    image: company-app:latest
    local: true
```
will mean that every build will push to the same tag, and broken developer builds could get mixed in with stable production builds. In this use case, teams will want to push to separate usernames/repos within the same local registry host, but use the same image name/tag scheme.

We could address this use case by adding another field - `prefixNamespace`, which would inject the current namespace name into the resolved pull spec. For example - if the local registry hostname is `image-registry.local:5000`:

```
kind: Build
metadata:
  namespace: staging
  name: company-app
spec:
  ...
  output:
    image: company-app:v1.0.0
    local: true
    prefixNamespace: true
```

will resolve the output image to `image-registry.local:5000/staging/company-app:v1.0.0`.

Reply:
As an aside, this is (sort of) how `ImageStreams` in OpenShift work. If an `ImageStream` is created in a namespace, the OpenShift image registry creates an image repo at `image-registry.openshift-image-registry.svc:5000/<namespace>/<name>`.

Reply:
This is a valid use-case but I would prefer to address this as a different issue, once the initial development for this proposal is finished.  In the other issue we could discuss all details around ` 1. How can we make `Build` objects with local image pull specs portable across namespaces? ` .So that we can merge this asap.